rust_src/src/multibyte.rs

```diff
@@ -106,6 +106,44 @@ impl LispStringRef {
     }
 }
 
+pub struct LispStringRefIterator<'a> {
+    string_ref: &'a LispStringRef,
+    cur: usize
+}
+
+// Substitue for FETCH_STRING_CHAR_ADVANCE
+impl<'a> Iterator for LispStringRefIterator<'a> {
+    type Item = (Codepoint, usize);
+
+    fn next(&mut self) -> Option<(Codepoint, usize)> {
+        if self.cur < self.string_ref.len_bytes() as usize {
+            let codepoint: Codepoint;
+            let ref_slice = self.string_ref.as_slice();
+            if self.string_ref.is_multibyte() {
+                let (cp, advance) = multibyte_char_at(&ref_slice[self.cur..]);
+                codepoint = cp;
+                self.cur += advance;
+            } else {
+                codepoint = ref_slice[self.cur] as u32;
+                self.cur += 1;
+            }
+            
+            Some((codepoint, self.cur))
+        } else {
+            None
+        }
+    }
+}
+
+impl LispStringRef {
+    pub fn iter(&self) -> LispStringRefIterator {
```

Comment:
Depending on need by other APIs, I think two iterators a la `chars()` and `char_indices()` would make sense.